chore: add info about upgrades. (#4188)
timescale/docs: use-timescale/upgrades.md

# Maintenance and upgrades

> $SERVICE_LONGs are managed cloud services offered by $COMPANY. To provide a stable and reliable environment for your 

Reviewer: Services are services?

Author: Who does not love a truism. Updated.

applications, each $SERVICE_SHORT is based on a specific version of the $PG database and the $TIMESCALE_DB extension.
To ensure that you benefit from the latest features, performance and security improvements, it is important that your

> $SERVICE_LONG is kept up to date with the latest versions $TIMESCALE_DB and $PG.

Reviewer: ```suggestion
$SERVICE_LONG is kept up to date with the latest versions of $TIMESCALE_DB and $PG.
```

$CLOUD_LONG has the following upgrade policies:
* **Minor software upgrades**: handled automatically, you do not need to do anything. 

  Upgrades are performed on your $SERVICE_LONG during a maintenance window that you 
  [define to suit your workload][define-maintenence-window]. You can also 
  [manually upgrade $TIMESCALE_DB][minor-manual-upgrade].

> * **Critical security upgrades**: are installed outside normal maintenance windows when necessary, and sometimes require 

Reviewer: ```suggestion
* **Critical security upgrades**: installed outside normal maintenance windows when necessary, and sometimes require 
```

  a short outage. 

>   Downtime is usually between 30 seconds and 5 minutes. $COMPANY endeavor to notify you by email 

Reviewer: ```suggestion
  Downtime is usually between 30 seconds and 5 minutes. $COMPANY aims to notify you by email 
```

  if downtime is required, so that you can plan accordingly. However, in some cases this is not possible.
* **Major upgrades**: such as a new version of $PG are performed [manually by you][manual-upgrade], or [automatically 
  by $CLOUD_LONG][automatic-upgrade]. 

<Highlight type="important">

After a maintenance upgrade, the DNS name remains the same. However, the IP address often changes.

</Highlight>

## Minor software upgrades

> If you do not [manually upgrade $TIMESCALE_DB][minor-manual-upgrade] for non-critical upgrades, upgrades are 

Reviewer: If ..., Timescale Cloud performs upgrades ...

Author: Done

Some minor upgrades do require some downtime. This is usually between 30 seconds and 5 minutes. If downtime is required 
for an upgrade, $COMPANY endeavors to notify you by email ahead of the upgrade. However, in some cases, we might not be 
able to do so. Best practice is to [schedule your maintenance window][define-maintenence-window] so that any downtime 
disrupts your workloads as little as possible and [minimize downtime with replicas][minimize-downtime]. If there are no 
pending upgrades available during a regular maintenance window, no changes are performed.

To track the status of maintenance events, see the Timescale [status page][status-page].

### Minimize downtime with replicas

Maintenance upgrades require up to two automatic failovers. Each failover takes less than a few seconds.
$SERVICE_LONGs with [$HA_REPLICAs and $READ_REPLICAs][replicas-docs] require minimal write downtime during maintenance, 
read-only queries keep working throughout. 

During a maintenance event, services with replicas perform maintenance on each node independently. When maintenance is 
complete on the primary node, it is restarted: 

> - If the restart takes more than a minute, the replica node is promoted to primary, given that the replica has no 

Reviewer: I assume something changes here for replica sets that have more than one node - need to clarify?

Author: Changed to say a node.

  replication lag. Maintenance now proceeds on the newly promoted replica, following the same
  sequence. If the newly promoted replica takes more than a minute to restart, the former
  primary is promoted back. In total, the process may result in up to two minutes of write
  downtime and two failover events. 
- If the maintenance on the primary node is completed within a minute and it comes back online, the replica remains
  the replica.

### Manually upgrade $TIMESCALE_DB for non-critical upgrades

Non-critical upgrades are available before the upgrade is performed automatically by $CLOUD_LONG. To upgrade 
$TIMESCALE_DB manually: 

<Procedure>

1. **Connect to your service**

   In [$CONSOLE][cloud-login], select the $SERVICE_SHORT you want to upgrade.

1. **Upgrade $TIMESCALE_DB**

   Either:
   - Click `SQL Editor`, then run `ALTEREXTENSION timescaledb UPDATE`. 

>    - Click the kebab menu, then `Pause` and `Resume` the service.

Reviewer: I don't think we use the name kebab menu anywhere, so I assume this should be the actual icon from the UI

Author: Found the icon, how cool.

</Highlight>

To upgrade your $SERVICE_SHORT to a newer version of $PG:

<Procedure>

1. **Connect to your service**

   In [$CONSOLE][cloud-login], select the $SERVICE_SHORT you want to upgrade.
1. **Disable $HA_REPLICAs**

   1. Click `Operations` > `High Availability`, then click `Change configuaration`.
   1. Select `Non-production  (No replica)`, then click `Change configuration`.

1. **Disable $READ_REPLICAs**

   1. Click `Operations` > `Read scaling`, then click the trash icon next to all replica sets.

1. **Upgrade $PG**
   1. Click `Operations` > `Service Upgrades`.
   1. Click `Upgrade service`, then confirm that you are ready to start the upgrade. 

   Your $SERVICE_LONG is unavailable until the upgrade is complete. This normally takes up to 20 minutes. 
   However, it can take longer if you have a large or complex $SERVICE_SHORT.

   When the upgrade is finished, your service automatically resumes normal
   operations. If the upgrade is unsuccessful, the service returns to the state
   it was in before you started the upgrade.

1. **Enable $HA_REPLICAs and replace your $READ_REPLICAs**

</Procedure>

## Automatic PostgreSQL upgrades for a $SERVICE_SHORT

> If you do not manually upgrade your $SERVICE_SHORTs within the customer self-service upgrade window, $CLOUD_LONG 

Reviewer: The words "customer" and "self-service" seem a bit weird here - does this mean if YOU do not upgrade manually during YOUR closest maintenance window, then TSC sees it as the sign to upgrade automatically?

Author: Linked back to deprecation section.

performs an automatic upgrade. Automatic upgrades can result in downtime, best practice is to 
[manually upgrade your $SERVICE_SHORTs][manual-upgrade] during a low-traffic period for your application.

During an automatic upgrade:
1. Any configured [$HA_REPLICAs][hareplica] or [$READ_REPLICAs][readreplica] are temporarily removed.
1. The primary $SERVICE_SHORT is upgraded.

> 1. $HA_REPLICAs and $READ_REPLICAs are added back to the $SERVICE_SHORT.

Reviewer: ```suggestion
1. $HA_REPLICA_CAPs and $READ_REPLICAs are added back to the $SERVICE_SHORT.
```

  replication lag. Maintenance now proceeds on the newly promoted replica, following the same
  sequence. If the newly promoted replica takes more than a minute to restart, the former
  primary is promoted back. In total, the process may result in up to two minutes of write
  downtime and two failover events. 
- If the maintenance on the primary node is completed within a minute and it comes back online, the replica remains
  the replica.

### Manually upgrade $TIMESCALE_DB for non-critical upgrades

Non-critical upgrades are available before the upgrade is performed automatically by $CLOUD_LONG. To upgrade 
$TIMESCALE_DB manually: 

<Procedure>

1. **Connect to your service**

   In [$CONSOLE][cloud-login], select the $SERVICE_SHORT you want to upgrade.

1. **Upgrade $TIMESCALE_DB**

   Either:
   - Click `SQL Editor`, then run `ALTEREXTENSION timescaledb UPDATE`. 
   - Click `⋮`, then `Pause` and `Resume` the service.

</Procedure>

## Deprecations

To ensure you benefit from the latest features, optimal performance, enhanced security, and full compatibility 
with $TIMESCALE_DB, $CLOUD_LONG supports a defined set of $PG major versions. To reduce the maintenance burden and 

> continue providing a high-quality managed experience, as $PG and $TIMESCALE_DB evolve,$COMPANY periodically deprecates 

Reviewer: ```suggestion
continue providing a high-quality managed experience, as $PG and $TIMESCALE_DB evolve, $COMPANY periodically deprecates 
```

deprecation is as follows:
- **Deprecation notice period begins**: you receive email notification of the deprecation and the timeline for the 
  upgrade.
- **Customer self-service upgrade window**: best practice is to [manually upgrade to a new $PG version][manual-upgrade] in
  this time.
- **Automatic upgrade deadline**: $CLOUD_LONG performs an [automatic upgrade][automatic-upgrade] of your $SERVICE_SHORT.

## Manually upgrade PostgreSQL for a $SERVICE_SHORT

Upgrading to a newer version of $PG enables you to take advantage of new features, enhancements, and security fixes. 
It also ensures that you are using a version of $PG that's compatible with the newest version of $TIMESCALE_DB. 

For a smooth upgrade experience, make sure you:

*  **Plan ahead**: upgrades cause downtime, so ideally perform an upgrade during a low traffic time.
*  **Run a test upgrade**: [fork your $SERVICE_SHORT][operations-forking], then try out the upgrade on the fork before 
   running it on your production system. This gives you a good idea of what happens during the upgrade, and how long it 
   might take. 

> *  **Keep a copy of your $SERVICE_SHORT**: if you're worried about losing tour data, 

Reviewer: ```suggestion
*  **Keep a copy of your $SERVICE_SHORT**: if you're worried about losing your data, 
```

   [fork your $SERVICE_SHORT][operations-forking] without upgrading, and keep this duplicate of your $SERVICE_SHORT. 
   To reduce cost, you can immediately pause this fork and only pay for storage until you are comfortable deleting it 
   after the upgrade is complete.

<Highlight type="important">

> Timescale services with replicas cannot be upgraded. To upgrade a service

Reviewer: ```suggestion
$SERVICE_LONGs with replicas cannot be upgraded. To upgrade a service
```

with a replica, you must first delete the replica and then upgrade the service.

</Highlight>

The following table shows you the compatible versions of $PG and $TIMESCALE_DB. 

<SupportMatrix />

For more information about feature changes between versions, see the
[PostgreSQL release notes][postgres-relnotes] and
[$TIMESCALE_DB release notes][timescale-relnotes].

<Highlight type="warning">

Your $SERVICE_LONG is unavailable until the upgrade is complete. This can take up to 20 minutes. Best practice is to 

> test on a fork first so you can estimate how long the upgrade will take. 

Reviewer: ```suggestion
test on a fork first, so you can estimate how long the upgrade will take. 
```
